### PR TITLE
Android: Replace movement buttons with joystick

### PR DESCRIPTION
**Pull request by srifqi**
Replace movement control buttons (arrow at bottom left screen) with joystick-like behavior.

Joystick has 8 directions (same as keyboard). Basically, just map it to keyboard input.
Joystick applies only on left 1/3 of screen. Joystick's position can be fixed by enabling `fixed_virtual_joystick` setting.

Three new images:
(1) placeholder joystick,
(2) joystick container (background), and
(3) joystick cursor.
Remove unused images: movement control buttons (*_arrow.png).

screenshot showing multi-touch ability
^ joystick + turn around + sneak (from video)

how it looks from a tablet device
^ How fixed joystick looks from a tablet device. Note that the joystick is **not** placed on left one-sixth and bottom one-third of screen. Also, fixed joystick has larger size when in standby.

[Joystick on standby (old)](https://user-images.githubusercontent.com/4017302/37554340-a1860ed4-2a09-11e8-8650-6483caeaf848.png)

[Concept (old video)](https://youtu.be/5RlkvvRvI3c)

[Video (1)](https://youtu.be/8Rgk7FjNDv0) [Video (2)](https://youtu.be/z9x4qkHGfrU)
[Video (3)](https://youtu.be/qp8XjzZ5t78) (fixed joystick's position)

### CODE REVIEW

**Comment by paramat** [comment]
The joystick pad image should be non-intrusive, mostly transparent.

How does Android currently do player turning? Screen-drag?

**Comment by srifqi** [comment]
Joystick image will be almost transparent, except when it is being used (like touched and dragged).

Yup! For Android, the screen split into two: left side for joystick and right side for turn around.

**Comment by srifqi** [comment]
I have added images as feedback for joystick input.

There are three images:
1. placeholder joystick,
2. joystick container (background), and
3. joystick cursor.

[Video (1)](https://youtu.be/8Rgk7FjNDv0) [Video (2)](https://youtu.be/z9x4qkHGfrU)

**Comment by stujones11** [comment]
I haven't has a chance to test this yet but it certainly looks to be an improvement on what we currently have and very similar to what I had in mind myself. I would, however, also like to get rid of those awful 'sneak' and 'jump' buttons and wonder if at least the jumping could be done by tapping the joystick. Perhaps adding a little inertial would prevent it from immediately snapping back to the centre when released.

I would also like that the controls are hidden if a hardware keyboard is connected, I am not sure if that is already the case or not, although it needn't be part of this PR either way.

I am a little concerned that the scale may be a bit much for larger screen devices but will tell you better when I try it on my tablet, maybe you could set a max and min size independent of screen dimension?

Also, I am not keen on those cursors, I do not think they help but that is just my personal preference.

**Comment by paramat** [comment]
Nes videos look good.
For less visual obstruction i think the joystick circles could both be completely transparent, the player will soon see how they work and won't need the slight shading.

**Comment by paramat** [comment]
On desktop both hands are needed movement too, left does WASD and right does turn, just like this PR.

**Comment by stujones11** [comment]
> On desktop both hands are needed movement too, left does WASD and right does turn

Yes, but the left hand also controls 'jump' & 'sneak', which gives PC users an unfair advantage. This is why the auto-jump feature was hard-coded for android, iirc again sapier :p

**Comment by paramat** [comment]
On this subject, part of the complaints about controls may well apply to the jump and sneak buttons, perhaps they can be spaced further apart, like for example move the jump button upwards so that it is diagonal to the sneak button and further from it, seems more natural for the thumb.
But this is probably for a separate PR.

**Comment by jastevenson303** [comment]
On that note: there is not a button for AUX1 (default [E]) on Android
either.

**Comment by srifqi** [comment]
The real joystick has back side which can be used for index finger and middle finger, but touch screen doesn't have that. The solution is to use two hands for player movement. Personally, I use three fingers: (1) left thumb for movement, (2) right thumb to jump/sneak, and (3) right index finger to turn around; as you can see from those videos.

I'll try to move the sneak button and jump button into a kind-of-circular position.

Keyboard hardware detection probably in another PR. :thinking:

About the transparency, IMO, we still need transparency as a sign for the player that it is being touched, as many other games do that. But, it can be just all transparent so we only need two images. :smile:

---

@jastevenson303 Maybe you want to go read Issue 2873? Maybe in another PR.

**Comment by srifqi** [comment]
The real joystick has back side which can be used for index finger and middle finger, but touch screen doesn't have that. The solution is to use two hands for player movement. Personally, I use three fingers: (1) left thumb for movement, (2) right thumb to jump/sneak, and (3) right index finger to turn around; as you can see from those videos.
As reference, touch controls in Minecraft: https://minecraft.gamepedia.com/Controls#Touch

I'll try to move the sneak button and jump button into a kind-of-circular position.

Keyboard hardware detection probably in another PR. :thinking:

About the transparency, IMO, we still need transparency as a sign for the player that it is being touched, as many other games do that. But, it can be just all transparent so we only need two images. :smile:

---

@jastevenson303 Maybe you want to go read Issue 2873? Maybe in another PR.

**Comment by srifqi** [comment]
first attempt to move jump button and sneak button
^ first attempt to move jump button and sneak button

---

Any ideas? ... or just move this in separate PR?

**Comment by stujones11** [comment]
> As reference, touch controls in Minecraft: https://minecraft.gamepedia.com/Controls#Touch

MCPE uses a D-pad with jump in the middle, only the 'sneak' button is at the far right. Using your thumb and index finger on a large screen tablet in not practical.

**Comment by paramat** [comment]
>  IMO, we still need transparency as a sign for the player that it is being touched, as many other games do that.

Well nevermind what other games do :) let's do what is logical and practical. How about the large circle is always completely transparent and the small inner circle is completely transparent when not touched and becomes shaded when touched? It will be obvious that the joystick is working due to player movement so i don't think the large circle needs to become shaded. I think it's more important to reduce the obscuring of the screen.
When not touching the joystick i think it's best both circles are completely transparent.

The jump / sneak buttons are probably best as a separate PR.

**Comment by rubenwardy** [comment]
I think the large circle should be shaded on press, to provide immediate feedback

**Comment by stujones11** [comment]
> Well nevermind what other games do :) let's do what is logical and practical

What we have now is neither logical nor practical, this (afaict) is merely 'prettying' things up a little, it does not make it any easier to play. Visually, it is as an improvement but I honestly don't think this is gonna wash with those that expressed dissatisfaction with the current controls.

And yes, we should look at how other games do these things, it needn't to be invented here :p

@paramat do you even play the mobile version of this game?

**Comment by rubenwardy** [comment]
It's very important what other games do, because players are used to it and expect it. It's a big part of usability. That doesn't mean what they do is the best way however

**Comment by stujones11** [comment]
Indeed, some people complain about the mcpe controls but you can't please everyone without providing a variety of options.

I personally find the old D-pad style control the most familiar and intuitive for this type of game.

My suggestion is to keep the two concentric circles of this design but keep the middle ring fixed as jump. The movement direction would be indicated by pointing arrows that highlight when you touch within the outer ring. Essentially a prettified and more modern looking  D-pad.

**Comment by paramat** [comment]
> I think the large circle should be shaded on press, to provide immediate feedback

Not my preference but that's ok with me.
>  this (afaict) is merely 'prettying' things up a little, it does not make it any easier to play. Visually, it is as an improvement but I honestly don't think this is gonna wash with those that expressed dissatisfaction with the current controls.

? This PR completely changes the WASD controls to what has been discovered to be popular with Android users and is agreed to be a good idea: a joystick reacting to direction of drag from initial touch point, see the videos to see the joystick reacting to different initial touch points.

Ok i should clarify, it is worth seeing what other games do, what i mean is not to give them too much importance if it is not necessary or not a good idea. See the context, the sentence below could be interpreted as meaning this was only being done because other games do, this was what i was reacting to, with a smiley to show it was meant with a sense of humour. Feedback may well be a good reason to do it.
> we still need transparency as a sign for the player that it is being touched, as many other games do that.

No i don't play touchscreen MT i wouldn't like touchscreen controls.

> My suggestion is to keep the two concentric circles idea of this design but have the middle ring fixed as jump. The movement direction would be indicated by pointing arrows that highlight when you touch within the outer ring. Essentially just a prettified and more modern looking D-pad.

Seems a bad idea. The movement controls would be very close to the jump button. Seems best that the controls are spaced out over the screen as in the PR, with joystick only doing movement.
Having the movement controls fixed is a disadvantage compared to a joystick that reacts the the point of inital press, this is the idea behind this PR.
Since Android users will usually be using autojump a jump button in the movement pad is somewhat unnecessary, it will get in the way and will be accidently triggered.

**Comment by stujones11** [comment]
> Since Android users will usually be using autojump

So does this mean it will remain hard-coded for android users?

IMO it needs to be if the jump button stays where it is now. Otherwise the game will remain unplayable on my device and I guess many others :(

Anyway, I'm gonna work on this myself so we will see what folks prefer ;)

**Comment by nerzhul** [comment]
i think two joystick would be better, as one is currently hidden (for player rotation), both should be in the bottom corners like a console, it's a great startup

File: src/gui/touchscreengui.cpp [hunk 1]
```diff
@@ -837,7 +870,74 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
...
+				if (dy < 0)
+					angle *= -1;
+				/* rotate to make comparing easier */
+				angle = fmod(angle + 180 + 22.5, 360);
+
+				/* reset state before applying */
+				for (unsigned int i = 0; i < 4; i ++)
+					m_joystick_status[i] = false;
+
+				if (distance <= g_settings->getU16("touchscreen_threshold")) {
```
Comment (nerzhul) [inline src/gui/touchscreengui.cpp]: cache this setting

File: src/gui/touchscreengui.cpp [hunk 2]
```diff
@@ -837,7 +870,74 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
...
+					// do nothing
+				} else if (angle < 45)
```
Comment (nerzhul) [inline src/gui/touchscreengui.cpp]: you should invert angle tests i think, because this matches for all others

File: src/gui/touchscreengui.cpp [hunk 3]
```diff
@@ -940,6 +1040,23 @@ bool TouchScreenGUI::doubleTapDetection()
 
 }
 
+void TouchScreenGUI::applyJoystickStatus()
+{
+	for (unsigned int i = 0; i < 4; i ++) {
+		SEvent translated;
+		memset(&translated, 0, sizeof(SEvent));
```
Comment (nerzhul) [inline src/gui/touchscreengui.cpp]: what ?
if you want to do that properly use C++ initializers

SEvent translated{};

File: src/gui/touchscreengui.cpp (hunk 2 above)
Comment (srifqi) [reply src/gui/touchscreengui.cpp]: I think after adding `angle` with 180 degree, it should not be negative number.

File: src/gui/touchscreengui.cpp (hunk 3 above)
Comment (srifqi) [reply src/gui/touchscreengui.cpp]: Oh, you're right. I just copied that function from another part of the code, seems like it hasn't been changed to C++ initializers.

File: src/gui/touchscreengui.cpp (hunk 2 above)
Comment (srifqi) [reply src/gui/touchscreengui.cpp]: Wait. Let me review my code.

**Comment by nerzhul** [comment]
Sorry i am tired. It's fine, i though these was negative numbers

**Comment by paramat** [comment]
> Then just review the code changes and stfu =)

stujones11 using an expletive in a insult against someone and then adding a smiley doesn't make it ok, it's obviously an attempt to get away with an insult by implying it is not serious. This behaviour is not acceptable here and if repeated can lead to repo ban, please be civil.
> do you even play the mobile version of this game?

Just because someone does not test this on Android does not make their opinions invalid. For some aspects of this PR it is not necessary to be testing this on Android. Of course i will want input from testers as i know i can't judge some other aspects.

The current hardcoded Android autojump is hacky will probably be removed but re-added in a different way. Autojump on Android is very popular so we intend to keep it in some form, the priority is the popularity of the official app.

I should add some detail here.

During discussions of improving Android controls, we considered what was popular with Android users, it was discovered that Multicraft's joystick that reacts to a drag from initial touch point was popular. This also removes problems with hitting exact locations and problems with multiple detection points close to each other. We came to some initial agreement to use this for our app, but of course we can consider alternatives.

The recent server stats show 90% is mobile and the most popular mobile app is Multicraft at 30%.
Multicraft will not be updating to MT 0.5.0 changes so it is likely there will be a very large number of mobile players wanting to move to a new app. The idea behind improving Android at this point is to bring those players into the official app, see #6731 Using Multicraft's joystick control is a good idea because of this.

stujones11 i think your approach may be good in some form and it's good to have an alternative, i just don't think the jump button should be in the WASD controls.

File: src/gui/touchscreengui.cpp (hunk 1 above)
Comment (srifqi) [reply src/gui/touchscreengui.cpp]: Will do.

**Comment by stujones11** [comment]
> stujones11 using an expletive

What expletive? You mean the one in the acronym that _you_ have determined to be one ;)

**Comment by paramat** [comment]
*sigh* =/

**Comment by stujones11** [comment]
@paramat it was never intended literally as an insult, I was just kinda saying you should reserve your opinions to the things you understand best.

Just for clarity, is wtf acceptable? or will that be a repo ban for me =)

**Comment by rubenwardy** [comment]
It's fine as it's not a personal attack.

I'd like to see the jump on the joystick tested, but maybe as a future PR.

**Comment by paramat** [comment]
Sorry to be clear, by 'if repeated' i mean if done many times, not one more time, we're quite lenient actually.

People are free to state their opinions on things outside their area of expertise.

**Comment by paramat** [comment]
Hmmm actually i can see advantages to fixed WASD buttons, and i was being too narrow minded earlier about what controls we will choose. Additional implementations are much appreciated.

**Comment by srifqi** [comment]
(Can we go back on-topic?)

Just pushed some fixes to nerzhul's review. Also, another things like dropping button's object after not being used, etc. Read the code. :smile:

Here is the APK file in case somebody interested to try:
https://drive.google.com/open?id=19AFd1O_0usEJmcI4khUYby8zlFflIf6e
(armeabi-v7a, 7bd1f994a4b5240d14e2192fbb32fc9d207a5c25, `minetest_game`)

Make sure you've deleted your Minetest folder (`/sdcard/Minetest/`) or just rename it or move it somewhere else.

---

BTW, should I fix that LINT error?

**Comment by IhrFussel** [comment]
I tested the linked .apk and it works well on Android 7.0

**Comment by srifqi** [comment]
I've made the unpressed joystick smaller.

Rebased to 5e61f64ce259fe0b23cbb377b44e90a0fbc820d7 and squashed.

smaller unpressed joystick

**Comment by srifqi** [comment]
Just fixing minor issues, rebased to recent `master` branch.

**Comment by MoNTE48** [comment]
@paramat I never said that MultiCraft will not upgrade to minetest 0.5. I still plan to do it "someday in the future." I'm glad that someone besides me is interested in the development of Minetest on mobile platforms. I will definitely test all your changes and maybe I will be able to offer something interesting in the future.

**Comment by paramat** [comment]
Ok i misunderstood.

**Comment by SmallJoker** [comment]
I successfully tested the latest apk in singleplayer mode on salted Android 6.0. The movement wheel is already a large improvement, but in my opinion the jump and sneak buttons should be larger to make it easier to trigger them.

**Comment by Wayward1** [comment]
This is great! It's definitely a huge improvement over the current controls :)

**Comment by paramat** [comment]
> the jump and sneak buttons should be larger (as in: higher) to make it easier to trigger them.

Because the player's thumbs are likely to be at equal heights?

I think the jump button should be an upward arrow since it is also used for up on ladders and with fly. The current jump icon doesn't make much sense and is too dense, it obscures the world behind it.
Simply flipping the down arrow would be good. Then there is possibility of enlarging the icons without obscuring the world too much.
Also, the up/down icons can probably have thinner lines to be more consistent with the joystick and obscure even less.
I can work on the icon images if you prefer not to.

**Comment by Wayward1** [comment]
> Is player view up/down done by screen drag on right side of screen as with view left/right?

Yes, the right side of the screen gives the same view mobility as a mouse/touchpad

**Comment by rubenwardy** [comment]
Let's not get side tracked, it's best if this PR only removes the directions and adds the joysticks. Other buttons can wait to other PRs

File: src/gui/touchscreengui.cpp [hunk 4]
```diff
@@ -768,13 +784,35 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
...
 			m_settingsbar.deactivate();
 			m_rarecontrolsbar.deactivate();
-			/* if we don't already have a moving point make this the moving one */
-			if (m_move_id == -1) {
-				m_move_id                  = event.TouchInput.ID;
-				m_move_has_really_moved    = false;
-				m_move_downtime            = porting::getTimeMs();
-				m_move_downlocation        = v2s32(event.TouchInput.X, event.TouchInput.Y);
-				m_move_sent_as_mouse_event = false;
+			/* check where it starts */
```
Comment (rubenwardy) [inline src/gui/touchscreengui.cpp]: `// Select joystick when left 1/3 of screen dragged`

**Comment by rubenwardy** [comment]
Also, does anyone know how to get rid of that F5 text?

**Comment by Wayward1** [comment]
@rubenwardy: 
screenshot_2018-03-23-23-39-31

**Comment by rubenwardy** [comment]
I did think I was missing something. Oh dear XD

**Comment by Wayward1** [comment]
Lol xD in that case: 
screenshot_2018-03-23-23-47-55

**Comment by LoneWolfHT** [comment]
Why not add a setting that allows users to choose how their screen looks? They could select this Pull Request's method, the current one, or other method(s) that may be added in the future

**Comment by paramat** [comment]
> Might be worth experimenting with a fixed joystick center - ie: you can tap the edge to sharply move forward

That was the advantage of a fixed joystick (stujones11 approach) i realised, however the joystick reacting from initial touch position has it's own advantages.

**Comment by srifqi** [comment]
Thanks for all your suggestions and ideas!

I've added fixed joystick's position, you can use it by enabling `fixed_virtual_joystick` setting (Controls section). When enabled, joystick control starting point is limited to unpressed joystick image instead of left 1/3 of screen.

See it in action: [Video (3)](https://youtu.be/qp8XjzZ5t78)

APK (armeabi-v7a, 95c0927, minetest_game):
_compiling (will upload soon)_

New in 97cd65f3793df9637e0c58c10392ada791505a19: some code cleanup and code style fixes
Rebased and squashed.

File: src/gui/touchscreengui.cpp [hunk 5]
```diff
@@ -415,6 +419,8 @@ TouchScreenGUI::TouchScreenGUI(IrrlichtDevice *device, IEventReceiver* receiver)
 		m_buttons[i].repeatdelay   = BUTTON_REPEAT_DELAY;
 	}
 
+	m_touchscreen_threshold = g_settings->getU16("touchscreen_threshold");
```
Comment (SmallJoker) [inline src/gui/touchscreengui.cpp]: This setting needs documentation

File: src/gui/touchscreengui.cpp [hunk 6]
```diff
@@ -837,7 +887,88 @@ void TouchScreenGUI::translateEvent(const SEvent &event)
...
+
+			if (m_joystick_has_really_moved ||
+					(!m_joystick_has_really_moved && inside_joystick) ||
+					(!m_fixed_joystick &&
+					distance_sq > m_touchscreen_threshold * m_touchscreen_threshold)) {
+				m_joystick_has_really_moved = true;
+				s32 distance = sqrt(distance_sq);
+
+				// angle in degrees
+				double angle = acos(dx / distance) * 180 / 3.1415926536;
```
Comment (SmallJoker) [inline src/gui/touchscreengui.cpp]: `M_PI` (`math.h`) is more precise and used a few times in other code.

File: src/gui/touchscreengui.cpp [hunk 7]
```diff
@@ -949,18 +1096,33 @@ TouchScreenGUI::~TouchScreenGUI()
 			btn->guibutton = NULL;
 		}
 	}
+
+	if (m_joystick_btn_off->guibutton != 0) {
```
Comment (SmallJoker) [inline src/gui/touchscreengui.cpp]: `if (m_joystick_btn_off->guibutton == nullptr) {` or simpler: `if (!m_joystick_btn_off->guibutton) {`

File: src/gui/touchscreengui.h [hunk 8]
```diff
@@ -201,68 +213,83 @@ class TouchScreenGUI
 	bool m_move_sent_as_mouse_event = false;
 	v2s32 m_move_downlocation = v2s32(-10000, -10000);
 
+	int m_joystick_id = -1;
+	bool m_joystick_has_really_moved = false;
+	v2s32 m_joystick_downlocation = v2s32(-10000, -10000);
+	bool m_fixed_joystick = false;
+	button_info *m_joystick_btn_off;
+	button_info *m_joystick_btn_bg;
+	button_info *m_joystick_btn_center;
```
Comment (SmallJoker) [inline src/gui/touchscreengui.h]: `= nullptr` for each. They're not initialized in the constructor.

**Comment by paramat** [comment]
Thanks for the option, options may be good to have, but even if not used at least we now have 2 implementations to try. Much appreciated.

**Comment by fozolo** [comment]
I know I'm late to this party but what if the joy stick could enable run when the user touches further way from the center?  So normal walk as is, but if they drag the joystick beyond the current outer radius, then run in that direction.

**Comment by srifqi** [comment]
@fozolo how we want to "sprint"? Are we going to use use/sprint key (<kbd>E</kbd>) like some mods are using? Remember that we actually just map joystick's direction to keyboard input.

**Comment by fozolo** [comment]
@srifqi yes exactly.

Dragging beyond the normal joystick radius could be like pressing that direction and <kbd>E</kbd> simultaneously.  This seems like a simple way for mobile users to utilize fast mode provided they have that privilege.

**Comment by paramat** [comment]
From the fixed joystick video it looks like you can tap the edge of the circle for fine movements, good. However the area that can be tapped seems too small to be able to reliably tap one direction out of 8 with a thumb. The area seems to need to be larger than the 'standby' joystick area. It's a little hard to judge how big this is though from a video.

I don't see any need for the large joystick circle to be shaded when touched, this obscures the world and the way the large circle gets bigger on touch is enough feedback.

**Comment by paramat** [comment]
E is often mapped to other functions though and you wouldn't want to accidently enable it using the joystick. Also, to use E without moving then becomes impossible.
If E is added to Android controls i think it is best a separate button on the left side above the joystick.

**Comment by paramat** [comment]
Here's a much better texture for jump / up / climb:

jump_btn

**Comment by bendeutsch** [comment]
Am I reading the code correctly that this still emulates four keyboard directions, i.e. a "DPad"? Minetest has analogue joystick support, doesn't it? What about emulating that instead? Or is that out of scope for this PR to keep the changes small?

**Comment by stujones11** [comment]
True analogue controls would not work well with the current movement code and may even provide an unfair advantage. IMO the android controls should be as close as possible to those of a PC/laptop. I still think that the virtual D-pad + touchpad method used by MCPE is about as close as it gets.

**Comment by rubenwardy** [comment]
Android users are *already* at a major disadvantage, so I don't see any issues with a continuous joystick

**Comment by stujones11** [comment]
> Android users are already at a major disadvantage

[citation needed]

If you are playing on your wristwatch then I am sure it will be a bit more challenging  :p)

**Comment by rubenwardy** [comment]
See CTF. Android users are at a major disadvantage and this is due to movement and other controls

The highest scoring Android player is at #80 and has a kills/deaths of 0.4

**Comment by stujones11** [comment]
I get that the current controls are bad, I was only trying to point out that not all Android users are necessarily crappy cell-phone users.

Are you sure the disadvantages on CTF are really about movement or could there be other factors involved there?

**Comment by paramat** [comment]
>  IMO the android controls should be as close as practically possible to those of a PC/laptop.

I agree, not necessarily for 'advantage' reasons, but more for controls similarity with desktop, and simplicity.

**Comment by stujones11** [comment]
> Also, controller support already has this

If the PC version already has analogue control via usb gamepad or whatever then I retract my remark about giving an unfair advantage.

_If_ that is the case then this definitely should mimic that instead of the keyboard.

**Comment by srifqi** [comment]
I've used offset instead of percentage to place joystick. Now, it is spaced one button size from bottom and from left.

how it looks from a tablet device
^ How it looks from a tablet device. Note that the joystick is not placed on left one-sixth and bottom one-third of screen.

File: src/gui/touchscreengui.cpp (hunk 5 above)
Comment (srifqi) [reply src/gui/touchscreengui.cpp]: Done.

**Comment by paramat** [comment]
I can't see how a continuous joystick makes controls easier.

Consider it has a disadvantage in that it is then almost impossible to walk exactly forward without slowly drifting to the sides, as there is no longer snappng to 8 directions. A touch control instead of a physical joystick makes it even harder to hit a precise direction.
As a player i would prefer to precisely walk in the direction i'm facing, for example fixing my crosshair on the end of a narrow path to avoid falling off the sides etc.
So i think it has to be an option alongside 8 directions.

The large majority of desktop users are using WASD, not a continous joystick, so i think it's best to be similar to WASD as default. There could be an option for continuous in a later PR.

Also consider we are trying to bring players in from forked apps, what are they using?

**Comment by rubenwardy** [comment]
>  it is then almost impossible to walk exactly forwar

There's this then called dampening. If it's close enough then it can just snap

-------

If you look above you can see I approved the PR, so I'm fine with this PR in its current state. Improvements can come later

**Comment by srifqi** [comment]
@paramat As already suggested by I-forget-who-was-it, I already set the fixed joystick size in "standby" mode to be the same size with "pressed" mode.

https://github.com/srifqi/minetest/blob/93b37683eeac806548a7a9b2c46b67000cb0b279/src/gui/touchscreengui.cpp#L479-L494

**Comment by paramat** [comment]
Ok, thanks.

**Comment by paramat** [comment]
See #2456 perhaps joystick needs a drop-shadow?

**Comment by LoneWolfHT** [comment]
How do I use this on my mobile device? I find it very hard to move around when using the dpad. Joysticks are a lot easier to use

**Comment by paramat** [comment]
Gimp has 'drop shadow' under filters.

**Comment by paramat** [comment]
Yes i'm fine with a merge.

**Comment by srifqi** [comment]
Thank you for the merge!

Here is the APK form recent master branch just in case somebody wants to try.
APK (armeabi-v7a, 326eeca306f7bfb53ae3685eef18978dd81e587e, minetest_game):
https://drive.google.com/open?id=1jBaudKOPN1l1qxVIqdSBaDODyTuvHctg

**Pull request merged**